server/src/main/java/io/deephaven/server/plugin/type/JsPluginDistribution.java

```diff
@@ -0,0 +1,137 @@
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.Objects;
+
+/**
+ * A {@link JsPlugin} implementation sourced from a distribution directory.
+ */
+public final class JsPluginDistribution extends JsPluginBase {
+
+    private static final String PACKAGE_JSON = "package.json";
+
+    /**
+     * Creates a new js plugin distribution. Assumes that {@value PACKAGE_JSON} exists in {@code distributionDir}. The
+     * {@link #name() name}, {@link #version() version}, and {@link #main() main} from {@value PACKAGE_JSON} will be
+     * used.
+     *
+     * @param distributionDir the distribution directory
+     * @return the js plugin distribution
+     * @throws IOException if an I/O exception occurs
+     */
+    public static JsPluginDistribution fromDistribution(Path distributionDir) throws IOException {
+        final Path packageJson = distributionDir.resolve(PACKAGE_JSON);
+        try (final InputStream in = new BufferedInputStream(Files.newInputStream(packageJson))) {
+            final PackageJson p = new ObjectMapper()
+                    .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+                    .readValue(in, PackageJson.class);
+            // Note: we could provide an option, or do it by default, to only copy over p.main, or p.main.parent, if we
+            // wanted to minimize the amount of data we re-expose.
+            return new JsPluginDistribution(distributionDir, p.name, p.version, p.main);
+        }
+    }
+
+    private final Path distributionDir;
+    private final String name;
+    private final String version;
+    private final String main;
+
+    /**
+     * Creates a new js plugin distribution.
+     *
+     * <p>
+     * Note: unlike {@link #fromDistribution(Path)}, {@code distributionDir} does not need to contain
+     * {@value PACKAGE_JSON}.
+     *
+     * @param distributionDir the distribution directory
+     * @param name the name, see {@link JsPlugin#name()}
+     * @param version the version, see {@link JsPlugin#version()}
+     * @param main the main, see {@link JsPlugin#main()}
+     */
+    public JsPluginDistribution(Path distributionDir, String name, String version, String main) {
+        this.distributionDir = Objects.requireNonNull(distributionDir);
+        this.name = Objects.requireNonNull(name);
+        this.version = Objects.requireNonNull(version);
+        this.main = Objects.requireNonNull(main);
+    }
+
+    @Override
+    public String name() {
+        return name;
+    }
+
+    @Override
+    public String version() {
+        return version;
+    }
+
+    @Override
+    public String main() {
+        return main;
+    }
+
+    @Override
+    public void copyTo(Path destination) throws IOException {
+        copyRecursive(distributionDir, destination);
+    }
+
+    private static void copyRecursive(Path src, Path dst) throws IOException {
```

Comment:
The javadocs for copy say that it follows symlinks - if you _don't_ want that behavior, we can use `NOFOLLOW_LINKS`.

I'll add in `COPY_ATTRIBUTES`.

Reply:
Yes, I'm mixing read and write - and if jimfs isn't the only possible fs impl, we need to be more careful of what could happen on writes, guard against files (or directories) already existing, not being empty, etc. Or recursively delete ahead of time?

Reply:
ATM, we are guarded against overwriting since we can't register the same plugin name multiple times.

Reply:
Guarded by implementation only - should probably have an explicit check for it, since the impl will happen to fail, as opposed to this not being a real use case (and indeed this _is_ a real use case, we do intend to support it).

Reply:
> I don't understand Q.

A user attempting to runtime-register a plugin will see it succeed. But if they try to update it and re-register it, it will fail, and not with a "this isn't support" error, instead they'll get a "IOException: file exists" error, probably with the "absolute" path within the zip (easily confused for their own path). No amount of trying to delete their own file so that it can be created will make sense for them. Instead, if it doesn't work, and it doesn't work _by design_, we should give them a coherent error around that, rather than just getting lucky that it fails, and leaving them to figure out why on their own.